### ISSUE TITLE
Siege engine issue

### ISSUE THREAD

**Issue by fricy** [mifki/df-twbt#21]
Referencing bug report: https://github.com/DFHack/dfhack/issues/408

**Comment by lethosor** [comment]
From what I understand, the problem here is that TwbT can't handle other code drawing to the map area. Would the ability to "hook" into `Screen::paintTile()` be useful?

**Comment by pronvit** [comment]
@lethosor well, it's simple - there are now two separate buffers for map area and for text, with two different dimensions. So plugins must draw into the right buffer and use the right dimensions. Because there's no way to automatically determine whether plugins wants to modify the map area or output some text over the map. Ideally TwbT should be recognised by dfhack drawing functions and all plugins updated upstream where necessary. For lua gui scripts I don't think there's even a way to fix them apart from supporting TwbT in dfhack.

Maybe it's time to do that already since TwbT is quite stable and widely used at least in the fortress mode and it's not good that it's making some plugins not work.

In my mousequery plugin there's also code to allow map dragging with right mouse button, so not all the changes are related to TwbT.

**Comment by lethosor** [comment]
My suggestion was a way for TwbT to intercept calls to `Screen::paintTile()` (which should cover all drawing functions, even from Lua). I assume TwbT itself would be able to determine if a plugin is drawing on the map or creating a viewscreen simply by determining the current viewscreen and the screen coordinates the plugin is drawing to (i.e. drawing to the map buffer if the coordinates fall within it).
It would be best not to integrate plugins into the DFHack core, not to mention the known issues TwbT causes (with adventure mode, Lua GUI scripts, rendermax, etc.). One solution could be to implement a way of hooking into core functions, which could be expanded to support the functions that your version of the mousequery plugin alters (Screen::paintTile, Gui::getDwarfmodeViewDims, etc.), although I'm not sure how to handle these sections: [1](https://github.com/mifki/df-twbt/blob/master/plugins/mousequery.cpp#L73-L77), [2](https://github.com/mifki/df-twbt/blob/master/plugins/mousequery.cpp#L648-L658), [3](https://github.com/mifki/df-twbt/blob/master/plugins/mousequery.cpp#L316-L318).
Non-TwbT mousequery changes should probably be merged into DFHack/dfhack to ensure a consistent user experience. (It's worth noting that DFHack core now exposes the key modifier state, so implementing controls like shift+drag should be easier.)

**Comment by pronvit** [comment]
@lethosor No, it's not possible to determine which buffer to draw into. Map dimensions may be smaller or bigger than text dimensions depending on tile sizes, and given a coordinate only, it's impossible to determine is it on the map or not. Also there definitely has to be two functions to get size of the map area and of the text screen.

I'd suggest adding a parameter something like useMapBuffer to drawing functions and functions returning dimensions and default it to false because most plugins output text and won't need changes.

**Comment by lethosor** [comment]
In that case, I think that exporting functions would be easier to handle than hooks. I've managed to come up with a way for plugins to expose functions to each other in [this branch](https://github.com/lethosor/dfhack/tree/plugin-exports). This would effectively allow plugins to do certain things when TwbT is available with:
```cpp
#include "twbt.h"
auto twbt = GET_PLUGIN_EXPORTS("twbt", twbt_export_class);
if (twbt)
  mpos.z += twbt->depth_at(mx, my);
```
This is already possible from Lua scripts (and the same functions can be exposed to C++ and Lua with minimal effort), although I'm not aware of an easy way to detect if a plugin is available and/or loaded.

**Comment by pronvit** [comment]
But that's almost how it works currently. Renderer object is always accessible, so you call enabler->renderer->is_twbt() that's inlined and checks for a magic value in renderer object. Not as generic as your mechanism but works and also tells if twbt is actually the renderer and not just loaded.

This doesn't solve the problem with determining map area size and drawing into the correct buffer though.

And getting tile depth is minor and likely needed only for mousequery.

**Comment by lethosor** [comment]
> This doesn't solve the problem with determining map area size and drawing into the correct buffer though.
I'm not sure what you mean. It should work the way it does in your versions of these plugins, with `enabler->renderer->is_twbt()` replaced with `twbt` (something like `twbt && twbt->enabled()` could be used to detect whether the TwbT renderer is actually in use). 

Aside from being generic and easily usable by other plugins, this is less dependent on plugin internals than things like `renderer_cool *r = (renderer_cool*)enabler->renderer` (which requires the `renderer_cool` class to be available to other plugins, which isn't really necessary for the few things it's used for). It's also easier to make these functions available to lua scripts that need them, rather than having different ways to do the same thing from C++ and Lua.

(The Lua-related problem I mentioned above is that mkmodule() calls open_plugin(), which produces an error for nonexistent plugins. It should be fairly easy to work around and provide a fallback `isEnabled` or `inUse` function.)

**Comment by pronvit** [comment]
> I'm not sure what you mean. It should work the way it does in your versions of these plugins, with enabler->renderer->is_twbt() replaced with twbt

So you're proposing to add such checks to each affected plugin in dfhack? How is this better than what I suggested before - "adding a parameter something like useMapBuffer to drawing functions and functions returning dimensions and default it to false because most plugins output text and won't need changes".

> which requires the renderer_cool class to be available to other plugins, which isn't really necessary for the few things it's used of

Didn't understand this. How were you going to call `twbt->enabled()` or `twbt->depth_at()` if `twbt` class is not available? The difference is just how you get this object - via `enabler->renderer` or `GET_PLUGIN_EXPORTS`, isn't it?

**Comment by lethosor** [comment]
I don't have a problem with adding a useMapBuffer parameter, but that won't work for everything (see the three examples I linked above) as there is currently no core function that provides the behavior of depth_at or [this](https://github.com/mifki/df-twbt/blob/master/plugins/mousequery.cpp#L648-L658).

I probably should have provided an example, so [here](https://gist.github.com/lethosor/05eac442ee71d478b980) are two (very rough) example plugins (the first plugin calls `dostuff` in the second). If the second plugin is not available or loaded, GET_PLUGIN_EXPORTS will return a null pointer. The class used by both plugins is in [this file](https://gist.github.com/lethosor/05eac442ee71d478b980#file-testplug2-h), which I admit isn't the most elegant solution, but it's simpler (and more similar to Lua) to pass around a single pointer than to locate individual function pointers and typecast them when necessary. If you have any suggestions to make this easier to maintain, I'd be happy to do so.

**Comment by pronvit** [comment]
I'll take a look once I'm at home but generally that's fine then. Anyway, rendering and dimensions that affect many plugins (and this issue btw) are much more important than about depth_at() that's used in mousequery only.

**Comment by pronvit** [comment]
Oh so testplugin2.h should better be called testplugin2_exports.h because it's not the same as testplugin2's internal .h but rather it's public interface only, right? That's what confused me when you used twbt.h in your code above - I thought it's just the same whole renderer class.
I'm not a C++ expert, can't tell if there's any cplusplusy way to get rid of duplicate method declarations.

**Comment by lethosor** [comment]
DFHack/dfhack#746 is the preferred way to fix this now. It'll require a few changes to TwbT, as well as getting rid of the extra plugins in this repo and merging in any non-TwbT-related changes.

According to testing done by gromit190 on IRC, TwbT is interfering with gps.dimx and dimy, which effectively prevents any scripts from determining the window size when TwbT is active (getPanelLayout() is useless too).

**Comment by pronvit** [comment]
@lethosor, you haven't added the necessary fields to DwarfmodeDims structure. Depending on what they do, plugins may want to know the size of the map (in map tiles) or the location/size of the map, area map and menu in text tiles. Plugins will need to be updated to use the correct values, but without the fields I can't even implement the hooks in TWBT.

**Comment by lethosor** [comment]
Progress!
![image](https://cloud.githubusercontent.com/assets/3719547/26641502/d0cf7428-45f8-11e7-8554-958fbb0eda57.png)
